base/float.jl

```diff
@@ -414,9 +422,121 @@ muladd(x::T, y::T, z::T) where {T<:IEEEFloat} = muladd_float(x, y, z)
 # TODO: faster floating point fld?
 # TODO: faster floating point mod?
 
-rem(x::T, y::T) where {T<:IEEEFloat} = rem_float(x, y)
+function unbiased_exponent(x::T) where {T<:IEEEFloat}
+    exp = (reinterpret(Unsigned, x) & exponent_mask(T)) >> mantissa_width(T)
+    return exp
+end
+
+function explicit_mantissa_noinfnan(x::T) where {T<:IEEEFloat}
+    subnormal = !issubnormal(x) ? significand_mask(T) + uinttype(T)(1) : uinttype(T)(0)
+    return subnormal | (significand_mask(T) & reinterpret(Unsigned, x))
+end
+
+function make_value(number::T, ep) where {T<:Unsigned}
+    Tfloat = floattype(T)
+    Tint = signed(T)
+    epint = unsafe_trunc(Tint,ep)
+    lz::signed(T) = Core.Intrinsics.ctlz_int(number) - exponent_width(Tfloat)
+    number <<= lz
+    epint -= lz
+    bits = T(0)
+    if epint >= T(0)
+        mant = number & significand_mask(Tfloat) #this might not be necessary
+        bits |= mant
+        epint = ((epint + Tint(1)) << mantissa_width(Tfloat)) & exponent_mask(Tfloat)
+        bits |= epint
+    else
+        number = number >> -epint
+        mant = number & significand_mask(Tfloat)
+        bits |= mant
+    end
+    return reinterpret(Tfloat, bits)
+end
+
+function fmod_internal(x::T, y::T) where {T<:IEEEFloat}
+    xuint = reinterpret(Unsigned, x)
+    yuint = reinterpret(Unsigned, y)
+    if xuint <= yuint
+        if xuint < yuint
+            return x
+        end
+        return zero(T)
+    end
+
+    e_x = unbiased_exponent(x)
+    e_y = unbiased_exponent(y)
+    # Most common case where |y| is "very normal" and |x/y| < 2^EXPONENT_WIDTH
+    if e_y > mantissa_width(T) && (e_x - e_y) <= exponent_width(T)
+        m_x = explicit_mantissa_noinfnan(x)
+        m_y = explicit_mantissa_noinfnan(y)
+        d = (e_x == e_y) ? (m_x - m_y) : (m_x << (e_x - e_y)) % m_y
+        iszero(d) && return zero(T)
+        return make_value(d, e_y - uinttype(T)(1))
+    end
+    # Both are subnormals
+    if e_x == uinttype(T)(0) && e_y == uinttype(T)(0)
+        bits = uinttype(T)(0)
+        number = xuint % yuint
+        mant = number & significand_mask(T) #this might not be necessary
+        bits |= mant
+        return reinterpret(T, bits)
+    end
+
+    m_x = explicit_mantissa_noinfnan(x)
+    e_x -= uinttype(T)(1)
+    m_y = explicit_mantissa_noinfnan(y)
+    lz_m_y = exponent_width(T)
+    if e_y > uinttype(T)(0)
+        e_y -= uinttype(T)(1)
+    else
+        m_y = mantissa(y)
+        lz_m_y = Core.Intrinsics.ctlz_int(m_y)
+    end
+
+    tz_m_y = Core.Intrinsics.cttz_int(m_y)
+    sides_zeroes_cnt = lz_m_y + tz_m_y
+
+    # n>0
+    exp_diff::uinttype(T) = e_x - e_y
```

Comment:
It'd be nice to determine whether/why these type assertions are necessary.

Reply:
Only one of them was